File: golang/internal/progress/progress.go
package progress

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"
)

type ProgressBar struct {
	total       int64
	completed   int64
	width       int
	startTime   time.Time
	description string
	mu          sync.Mutex
	started     bool
}

func NewProgressBar(total int64, description string) *ProgressBar {
	return &ProgressBar{
		total:       total,
		width:       40,
		description: description,
		startTime:   time.Now(),
	}
}

func (p *ProgressBar) Update(delta int64) {
	p.mu.Lock()
	p.completed += delta
	p.render()
	p.mu.Unlock()
}

func (p *ProgressBar) SetTotal(total int64) {
	p.mu.Lock()
	p.total = total
	p.mu.Unlock()
}

func (p *ProgressBar) SetDescription(desc string) {
	p.mu.Lock()
	p.description = desc
	p.mu.Unlock()
}

func (p *ProgressBar) SetCompleted(completed int64) {
	p.mu.Lock()
	p.completed = completed
	p.render()
	p.mu.Unlock()
}

func (p *ProgressBar) Finish() {
	p.mu.Lock()
	if !p.started {
		p.started = true
		p.startTime = time.Now()
	}
	p.completed = p.total
	p.render()
	fmt.Fprintln(os.Stderr)
	p.mu.Unlock()
}

func (p *ProgressBar) GetCompleted() int64 {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.completed
}

func (p *ProgressBar) render() {
	if p.total == 0 {
		return
	}

	percent := float64(p.completed) / float64(p.total)
	if percent > 1.0 {
		percent = 1.0
	}

	filled := int(percent * float64(p.width))

	elapsed := time.Since(p.startTime)
	_ = elapsed

	fmt.Fprintf(os.Stderr, "\r%s: [%s%s] %3.0f%% | %d/%d",
		p.description,
		strings.Repeat("=", filled),
		strings.Repeat(" ", p.width-filled),
		percent*100,
		p.completed,
		p.total)
}

type MultiProgress struct {
	bars     []*ProgressBar
	mu       sync.Mutex
	active   int
	maxDescr int
}

func NewMultiProgress() *MultiProgress {
	return &MultiProgress{
		maxDescr: 20,
	}
}

func (m *MultiProgress) AddBar(total int64, description string) *ProgressBar {
	m.mu.Lock()
	defer m.mu.Unlock()

	if len(description) > m.maxDescr {
		description = description[:m.maxDescr-3] + "..."
	} else if len(description) < m.maxDescr {
		description = strings.Repeat(" ", m.maxDescr-len(description)) + description
	}

	bar := &ProgressBar{
		total:       total,
		width:       30,
		description: description,
		startTime:   time.Now(),
		started:     true,
	}
	m.bars = append(m.bars, bar)
	m.active++
	return bar
}

func (m *MultiProgress) RemoveBar(bar *ProgressBar) {
	m.mu.Lock()
	defer m.mu.Unlock()

	for i, b := range m.bars {
		if b == bar {
			m.bars = append(m.bars[:i], m.bars[i+1:]...)
			m.active--
			break
		}
	}
}

func (m *MultiProgress) RenderAll() {
	m.mu.Lock()
	defer m.mu.Unlock()

	for _, bar := range m.bars {
		bar.render()
		fmt.Fprintln(os.Stderr)
	}
}

func GetCPUCount() int {
	return runtime.NumCPU()
}

func FormatNumber(n int64) string {
	if n >= 1_000_000_000 {
		return fmt.Sprintf("%.2fB", float64(n)/1_000_000_000)
	} else if n >= 1_000_000 {
		return fmt.Sprintf("%.2fM", float64(n)/1_000_000)
	} else if n >= 1_000 {
		return fmt.Sprintf("%.2fK", float64(n)/1_000)
	}
	return fmt.Sprintf("%d", n)
}

File: rust-miller-rabin/src/progress.rs
//! Lightweight progress bar implementation for Miller-Rabin testing
//!
//! Uses ANSI escape codes with no external dependencies beyond the standard library.

use std::io::Write;
use std::sync::Mutex;
use std::time::{Duration, Instant};

/// Internal state for the progress bar
struct ProgressState {
    completed: usize,
    last_update: Instant,
}

/// A thread-safe progress bar for tracking Miller-Rabin test progress
///
/// This implementation uses a Mutex for thread safety and supports
/// updating from multiple threads during parallel execution.
///
/// # Examples
/// ```
/// use miller_rabin_tester::progress::ProgressBar;
/// use std::sync::Arc;
///
/// let progress = Arc::new(ProgressBar::new(100, "Testing primality"));
/// progress.update(10);
/// progress.update(20);
/// progress.finish();
/// ```
pub struct ProgressBar {
    total: usize,
    state: Mutex<ProgressState>,
    width: usize,
    description: String,
    start_time: Instant,
    update_interval: Duration,
}

impl ProgressBar {
    /// Creates a new progress bar with the given total steps and description
    ///
    /// # Arguments
    /// * `total` - The total number of steps to completion
    /// * `description` - A description shown before the progress bar
    pub fn new(total: usize, description: &str) -> Self {
        Self {
            total,
            state: Mutex::new(ProgressState {
                completed: 0,
                last_update: Instant::now(),
            }),
            width: 40,
            description: description.to_string(),
            start_time: Instant::now(),
            update_interval: Duration::from_millis(50),
        }
    }

    /// Updates the progress bar by adding 'delta' completed steps
    ///
    /// This method is thread-safe and can be called concurrently from multiple threads.
    pub fn update(&self, delta: usize) {
        let mut state = self.state.lock().unwrap();
        state.completed += delta;
        let completed = state.completed;

        let now = Instant::now();
        if now.duration_since(state.last_update) >= self.update_interval || completed >= self.total
        {
            state.last_update = now;
            drop(state);
            self.render(completed);
        }
    }

    /// Marks the progress bar as complete
    ///
    /// This renders the bar at 100% and prints a newline.
    pub fn finish(&self) {
        self.render(self.total);
        eprintln!();
    }

    fn render(&self, completed: usize) {
        if self.total == 0 {
            return;
        }

        let percent = (completed as f64 / self.total as f64).min(1.0);
        let filled = (percent * self.width as f64) as usize;

        let filled_str = "=".repeat(filled);
        let empty_str = " ".repeat(self.width.saturating_sub(filled));

        let elapsed = self.start_time.elapsed().as_secs_f64();
        let rate = if elapsed > 0.0 {
            completed as f64 / elapsed
        } else {
            0.0
        };
        let rate_str = format_rate(rate);

        eprint!(
            "\r{}: [{}{}] {:3.0}% | {}/{} | {}    ",
            self.description,
            filled_str,
            empty_str,
            percent * 100.0,
            completed,
            self.total,
            rate_str,
        );
        let _ = std::io::stderr().flush();
    }
}

fn format_rate(rate: f64) -> String {
    if rate >= 1_000_000.0 {
        format!("{:.1}M/s", rate / 1_000_000.0)
    } else if rate >= 1_000.0 {
        format!("{:.1}K/s", rate / 1_000.0)
    } else {
        format!("{:.0}/s", rate)
    }
}

/// Callback type for progress reporting during Miller-Rabin testing
///
/// The callback receives (current_progress, total_work) where progress
/// is typically measured in bit operations or bases tested.
pub type ProgressCallback = dyn Fn(usize, usize) + Send + Sync;

/// Creates a callback that drives a ProgressBar from (current, total) values.
///
/// The callback computes the delta from the last reported position and
/// forwards it to `ProgressBar::update`.
pub fn create_progress_callback(
    progress: std::sync::Arc<ProgressBar>,
    total: usize,
) -> impl Fn(usize, usize) + Send + Sync {
    let last = std::sync::atomic::AtomicUsize::new(0);
    move |current, _total| {
        // Scale current into the ProgressBar's total range
        if total == 0 {
            return;
        }
        let scaled = (current as u128 * progress.total as u128 / total as u128) as usize;
        let prev = last.swap(scaled, std::sync::atomic::Ordering::Relaxed);
        if scaled > prev {
            progress.update(scaled - prev);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_progress_bar_creation() {
        let pb = ProgressBar::new(100, "Test");
        // Just verify it doesn't panic
        pb.update(10);
        pb.finish();
    }

    #[test]
    fn test_format_rate() {
        assert_eq!(format_rate(0.5), "0/s");
        assert_eq!(format_rate(1000.0), "1.0K/s");
        assert_eq!(format_rate(1500000.0), "1.5M/s");
    }
}
